record/envelope.go

```diff
@@ -0,0 +1,196 @@
+package record
+
+import (
+	"bytes"
+	"errors"
+	"fmt"
+	"time"
+
+	pool "github.com/libp2p/go-buffer-pool"
+	"github.com/libp2p/go-libp2p-core/crypto"
+	pb "github.com/libp2p/go-libp2p-core/record/pb"
+
+	"github.com/gogo/protobuf/proto"
+	"github.com/multiformats/go-varint"
+)
+
+// SignedEnvelope contains an arbitrary []byte payload, signed by a libp2p peer.
+//
+// Envelopes are signed in the context of a particular "domain", which is a
+// string specified when creating and verifying the envelope. You must know the
+// domain string used to produce the envelope in order to verify the signature
+// and access the payload.
+type SignedEnvelope struct {
```

Comment:
Let's just simplify and call this `Envelope`. `SignedEnvelope` makes the assumption that this is the only thing we'll ever do with an envelope.